### DISCUSSION TITLE
Observing both a model and its related models

### DISCUSSION THREAD

**Discussion by samul-1** [NilCoalescing/djangochannelsrestframework#146] [General]
I have a consumer that inherits from ObserverModelInstanceMixin. The model I'm subscribing to has several relations with other models, including many-to-many and many-to-one.

I would like to get notified when one of the related objects change, when the m2m relations are updated, and when a related object gets created or deleted.

Is there a way to subscribe to both a model and all of its relations?

**Comment by hishnash** [comment]
Yer is you will need to create a custom `@model_observer` that observes these other models and uses groups based on the relationship

if you consider a regular related model eg

```python
class Classroom(models.Model):
     school = models.ForeignKey(School, related_name='children')
     ...
````

And you want to subscribe to a School and get notified whenever a classroom is added/removed then you can create a custom model observer that observes changes to `Classroom ` and groups these changes based on the school.  (there is an example of this here https://github.com/NilCoalescing/djangochannelsrestframework#subscribing-to-a-filtered-list-of-models).  In this example if a class room is re-asigned from school A to school B, anyone subscribed to school A gets a 'deleted' notification and anyone subscribed to school B gets a 'created' notification. 

For many to many relationships you may need to explicitly create the ManyToMany table in django https://docs.djangoproject.com/en/4.0/topics/db/models/#extra-fields-on-many-to-many-relationships then you can create a custom observer on that many to many table to track additions/removals of related objects.  

You could create a single subscribe action that subscribes to the model instance and to all of these relationships you will need to explicitly create these.  Having the option on a ObserverModelInstanceMixin to also include all related objects might be a good idea but I do not think it should be default. Might be nice if you could add an Feature request issue to the repo to track this as a possible future feature with some details on how you would expect to customise this if at all.

**Reply by hishnash** [reply]
One extra thing to note for ManyToMany relationships if you want to track updates to the related object (not just the many to many table) then you will need to ensure the related object always pre-fetches its many-to-many relatsionship table entires since the `groups_for_signal` method is called very often (even when you do not make any changes so it can track if to send `updated` or `deleted` messages)

**Reply by samul-1** [reply]
Thank you for the insights. Unfortunately, at this time I am not familiar enough with django channels rest to try and tackle that on my own, but I'll make sure to follow up on your advice and create a feature request issue later today.

In my use case, I'm using the observer consumer inside an editor in my app where multiple users can access the resource in read mode, but only one at a time can edit it. While the "readers" are watching, I'm updating their frontend by showing any changes made to the resource by the "writer". This is nice, but the experience breaks as soon as the writer touches any of the related objects that are also shown in the editor. Ideally, the consumer would also notify about changes to those.
